### PR TITLE
daos: copy hdf5 containers to and from POSIX filesystems

### PR DESCRIPTION
**Pull request by dsikich**
If daos-api=hdf5, then h5repack is invoked. This is for copying to/from DAOS HDF5 containers. If the daos-api flag is not set, then the default is to move hdf5 files to DAOS POSIX containers. Also, h5repack is serial, so if more than one rank is specified when an HDF5 container is the source or destination the program will exit with an error explaining only one process is supported currently.

Signed-off-by: Danielle Sikich <danielle.sikich@intel.com>

### CODE REVIEW

File: src/common/mfu_daos.c
Comment (daltonbohning) [inline src/common/mfu_daos.c:1432]: @dsikich Else case for when NEITHER the src or dst is HDF5?

File: src/common/mfu_daos.c
Comment (daltonbohning) [inline src/common/mfu_daos.c:1384]: @dsikich We also need to free args[2]

File: src/common/mfu_daos.c
Comment (daltonbohning) [inline src/common/mfu_daos.c:1398]: @dsikich We also need to free args[2]

File: src/common/mfu_daos.c
Comment (daltonbohning) [inline src/common/mfu_daos.c:1406]: @dsikich We also need to free args[2]

File: src/common/mfu_daos.c
Comment (daltonbohning) [inline src/common/mfu_daos.c:1419]: @dsikich We also need to free args[2]

File: src/common/mfu_daos.c
Comment (daltonbohning) [inline src/common/mfu_daos.c:1426]: @dsikich Could we instead just check `mfu_dst_file->type`?

File: src/common/mfu_daos.c
Comment (daltonbohning) [inline src/common/mfu_daos.c:1431]: @dsikich We also need to free args[2]

File: src/common/mfu_daos.c
Comment (daltonbohning) [inline src/common/mfu_daos.c:336]: @dsikich Just a note that we'll need to incorporate DAOS_PROP_CO_LAYOUT_HDF5 into the dest logic rework. I don't expect there to be any "conflicts", but some of the logic might need some adjustments.

**Comment by daltonbohning** [comment]
This PR should be merged after #457

File: src/common/mfu_daos.c
Comment (daltonbohning) [inline src/common/mfu_daos.c:1345]: @dsikich After the serialization changes, we can use
```suggestion
        args[0] = HDF5_BIN_DIR"/h5repack";
```

File: src/dcp/dcp.c
Comment (daltonbohning) [inline src/dcp/dcp.c:370]: @dsikich We should only call this if either the source or dst is HDF5. Otherwise, we'll never get to the POSIX or DAOS copies

File: src/common/mfu_daos.c
Comment (daltonbohning) [inline src/common/mfu_daos.c:1429]: @dsikich I think this is more explicit. What do you think?
```suggestion
            if (mfu_dst_file->type == HDF5) {
                MFU_LOG(MFU_LOG_INFO, "Successfully copied to DAOS HDF5 container: %s",
                        dst_cont_uuid);
            } else {
                MFU_LOG(MFU_LOG_INFO, "Successfully copied to HDF5 file: %s",
                        argpaths[1]);
            }
```

File: src/common/mfu_daos.c
Comment (daltonbohning) [inline src/common/mfu_daos.c:1431]: @dsikich
```suggestion
        mfu_free(&args[2]);
        mfu_free(&args[3]);
```

File: src/dcp/dcp.c
Comment (daltonbohning) [inline src/dcp/dcp.c:380]: @dsikich We should only call this if either the source or dst is HDF5. Otherwise, we'll never get to the POSIX or DAOS copies. Also, maybe we can simplifythe if/else statement.
```suggestion
    if (mfu_src_file->type == HDF5 || mfu_dst_file->type == HDF5) {
        rc = mfu_daos_hdf5_copy(mfu_src_file, mfu_dst_file, argpaths, daos_args);
        if (rc != 0) {
            rc = 1;
        }
        mfu_finalize();
        MPI_Finalize();
        return rc;
    }
```

**Review by daltonbohning — COMMENTED** [review]
@dsikich Updated my previous comments to use commit suggestions, at your discretion.

File: src/common/mfu_daos.c
Comment (adammoody) [inline src/common/mfu_daos.c:1376]: Just a heads up that some MPI libraries do a poor job supporting fork() calls.  That means some users may encounter problems if their MPI library or their system happens to be one that has trouble with fork.  On the first IBM BlueGene systems, fork() wasn't an option at all.  On Infiniband networks, users may need to ``export IBV_FORK_SAFE=1`` to get things to work properly.

I don't know of a good substitute that can be used instead.  Just something to be aware of.  Perhaps add a note in our documentation where it would come into play.

Comment (adammoody) [reply src/common/mfu_daos.c:1376]: Oh, if HDF offers an API/library for this, that might be better than invoking their binary via fork().

Comment (daltonbohning) [reply src/common/mfu_daos.c:1376]: This is good to know. Whatever we ultimately end up doing, I'll be sure to run tests with the different systems and configurations that DAOS supports.

File: src/common/mfu_daos.c
Comment (dsikich) [reply src/common/mfu_daos.c:336]: @daltonbohning I think that h5repack should handle creating the destination. I am about to test some of this but this should work once  
DAOS-6574 duns: some bug fixes and features to DUNS is merged into DAOS.

**Comment by dsikich** [comment]
Leaving a note here that the functionality in this PR is only supported if commit dc6306dbed9698d9f6e218c5b3ec15c7f9d9044c from the HDF5 DAOS VOL is present. Also, this PR needs commit 506a8bda1ce81dd039e10e809f716c7f2d544a9b for DAOS.

File: src/common/mfu_daos.c
Comment (dsikich) [reply src/common/mfu_daos.c:336]: @daltonbohning after testing, h5repack does not accept daos://$pool/ without a generated UUID, so I think we will just need to generate a UUID for the destination in this case like we do the other tools.

Comment (dsikich) [reply src/common/mfu_daos.c:336]: @daltonbohning but we do not need to create the container

Comment (daltonbohning) [reply src/common/mfu_daos.c:336]: @dsikich I imagine that's probably because duns_resolve_path currently fails for daos://pool.
I wonder if Mohamad has plans to support daos://pool after his DAOS PR? https://github.com/daos-stack/daos/pull/4798#event-4414016922
But I guess worst case, we can just call duns_resolve_path, check for missing container, generate uuid if needed, rebuild the daos:// path, and pass to h5repack

Comment (dsikich) [reply src/common/mfu_daos.c:336]: @daltonbohning that is the PR I was using to run the h5repack, so daos://$pool/ will not work for h5repack even using it I guess. So we will still need to generate the cont UUID to pass to h5repack, and that handles container creation.

Comment (daltonbohning) [reply src/common/mfu_daos.c:336]: @dsikich As a reference, I have similar logic for doing this in my PR:
https://github.com/hpc/mpifileutils/pull/468/commits/2cfc511d5c1bd185a2e415d826954cc1ab3991e9
(scroll to line ~420)

For now, I would say to just do what you need to and I can handle any conflicts on my end, since my PR doesn't contain "necessary" changes.

**Comment by dsikich** [comment]
@daltonbohning I ran this with the new changes to hdf5 and DAOS and it seems to be working. I removed any code that connected/created/cleaned up pools or containers, as h5repack should handle that. This can be merged after daos-stack/daos#4798 lands, which looks like it should be soon.

File: src/common/mfu_daos.c
Comment (dsikich) [reply src/common/mfu_daos.c:4808]: @daltonbohning yeah, good catch. I just meant to have this for the destination container.

**Review by daltonbohning — APPROVED** [review]
@dsikich Just minor things. Otherwise, LGTM.

File: src/common/mfu_daos.c
Comment (dsikich) [reply src/common/mfu_daos.c:4852]: @daltonbohning yes, I agree.

File: src/common/mfu_daos.c
Comment (adammoody) [inline src/common/mfu_daos.c:978]: stray spaces or tab here?

File: src/common/mfu_daos.c
Comment (adammoody) [reply src/common/mfu_daos.c:4937]: Yeah, I was thinking of how to simplify this too.  Another option is to just always strdup and always free which you can do so long as all pointers are initialize to NULL if not being used:
```
args[1] = strdup("--dst-vol-name=native");
...
mfu_free(&args[1]);
mfu_free(&args[2]);
mfu_free(&args[3]);
mfu_free(&args[4]);
```
That works, because this is valid:
```
args[4] = NULL;
mfu_free(&args[4]);
```

File: src/dcp/dcp.c
Comment (adammoody) [inline src/dcp/dcp.c:425]: Maybe a tab in here?

**Review by adammoody — APPROVED** [review]

**Comment by adammoody** [comment]
This looks good.  I made a couple comments, but nothing to hold up the merge.  Thanks.

File: src/common/mfu_daos.c
Comment (daltonbohning) [reply src/common/mfu_daos.c:4937]: Ah, yes. I think that's even simpler!

Comment (dsikich) [reply src/common/mfu_daos.c]: @adammoody @daltonbohning ok, I updated this

**Comment by dsikich** [comment]
> This looks good. I made a couple comments, but nothing to hold up the merge. Thanks.

@adammoody thanks!

**Pull request merged**
